### PR TITLE
183 local developer and data scientists access to cosmos db

### PR DESCRIPTION
**Pull request by jedheaj314**
# Resolves _#183_

## How is this addressed

- Terraform platform.tf has been changed to add new resource block to add developers and data scientists to cosmos db.
- Pending work:
       check connectivity from a local machine (cloning the Dash repository and filing out connection string)
       add any Ip exceptions etc. required to establish a connection

## Has the CIS document been consulted / updated?
N/A

- Are you adding new resources, or modifying existing ones? - Adding new
- Have settings been set in line with the CIS recommendations? N/A

### CODE REVIEW

File: apps/main.tf [hunk 1]
```diff
@@ -25,27 +25,27 @@ data "external" "github_access_token" {
...
-  environment                      = var.environment
-  accesses_real_data               = var.accesses_real_data
-  webapps_subnet_id                = var.serve_webapps_subnet_id
-  resource_group_name              = var.core_rg_name
-  location                         = var.core_rg_location
-  log_analytics_name               = var.core_log_analytics_name
-  app_service_plan_name            = var.serve_app_service_plan_name
-  acr_name                         = var.serve_acr_name
-  cosmos_account_name              = var.serve_cosmos_account_name
-  feature_store_db_name            = var.transform_feature_store_db_name
```
Comment (jedheaj314) [inline apps/main.tf:43]: Working unintended merge changes

File: apps/app/webapp.tf [hunk 2]
```diff
@@ -216,3 +216,19 @@ resource "azuread_group_member" "developers" {
...
+  account_name        = var.cosmos_account_name
+  role_definition_id  = azurerm_cosmosdb_sql_role_definition.webapp.id
+  principal_id        = var.data_scientists_ad_group_principal_id
+  scope               = "${data.azurerm_cosmosdb_account.state_store.id}/dbs/${azurerm_cosmosdb_sql_database.app.name}"
+}
+
+resource "azurerm_cosmosdb_sql_role_assignment" "developers" {
+  resource_group_name = var.resource_group_name
+  account_name        = var.cosmos_account_name
+  role_definition_id  = azurerm_cosmosdb_sql_role_definition.webapp.id
```
Comment (t-young31) [inline apps/app/webapp.tf:231]: would suggest renaming this if it's going to be assigned to people as well as the webapp. maybe `read_write` or something to that effect?

Comment (jedheaj314) [reply apps/app/webapp.tf:231]: Done e0a7515d93e9f2a95b8e72b4a14d128b0a307214
Renamed it to "${var.app_id}-CosmosSingleDB-ReadWrite

Comment (t-young31) [reply apps/app/webapp.tf:231]: sorry – not clear. I meant the tf resource i.e. `azurerm_cosmosdb_sql_role_definition.read_write.id`

File: apps/app/webapp.tf [hunk 3]
```diff
@@ -216,3 +216,19 @@ resource "azuread_group_member" "developers" {
   group_object_id  = var.developers_ad_group_principal_id
   member_object_id = azuread_group.developers_of_this_app.id
 }
+
+resource "azurerm_cosmosdb_sql_role_assignment" "data_scientists" {
```
[the remark this answers is not in the document]
Comment (jedheaj314) [reply apps/app/webapp.tf:220]: How do you mean, could you please elaborate?
My understanding is data scientists need to access so we add the data scientists to the AD group (somewhere else)
and
here I am assigning the role to the group of data scientists to be able to access the cosmos

[the remark this answers is not in the document]
Comment (t-young31) [reply apps/app/webapp.tf:220]: was wondering why data scientists need access to the app state store. if that's where model predictions get stored then makes sense (missed lots of this morning's meeting 😢 so not sure on that one  – @jjgriff93 were you there?)
***
the latter to be more DRY: `for_each = toset([var.developers_ad_group_principal_id,  var.data_scientists_ad_group_principal_id])`

[the remark this answers is not in the document]
Comment (jjgriff93) [reply apps/app/webapp.tf:220]: And yes agree `for_each` is the way if so :)

File: apps/app/variables.tf [hunk 4]
```diff
@@ -103,6 +103,10 @@ variable "developers_ad_group_principal_id" {
   type = string
 }
 
+variable "data_scientists_ad_group_principal_id" {
```
[the remark this answers is not in the document]
Comment (jjgriff93) [reply apps/app/variables.tf:106]: Yeah why not - if AJ doesn't mind doing it in this PR, as the groups are used to access things created in both transform and serve guess it makes sense. Some for the developers ad group

File: apps/app/webapp.tf [hunk 5]
```diff
@@ -216,3 +216,19 @@ resource "azuread_group_member" "developers" {
...
+  resource_group_name = var.resource_group_name
+  account_name        = var.cosmos_account_name
+  role_definition_id  = azurerm_cosmosdb_sql_role_definition.webapp.id
+  principal_id        = var.data_scientists_ad_group_principal_id
+  scope               = "${data.azurerm_cosmosdb_account.state_store.id}/dbs/${azurerm_cosmosdb_sql_database.app.name}"
+}
+
+resource "azurerm_cosmosdb_sql_role_assignment" "developers" {
```
[the remark this answers is not in the document]
Comment (jjgriff93) [reply apps/app/webapp.tf:228]: Yep, same for the SQL group as well. `count = var.accesses_real_data ? : 0 : 1

File: apps/main.tf [hunk 6]
```diff
@@ -25,27 +25,28 @@ data "external" "github_access_token" {
...
+  resource_group_name                   = var.core_rg_name
+  location                              = var.core_rg_location
+  log_analytics_name                    = var.core_log_analytics_name
+  app_service_plan_name                 = var.serve_app_service_plan_name
+  acr_name                              = var.serve_acr_name
+  cosmos_account_name                   = var.serve_cosmos_account_name
+  feature_store_db_name                 = var.transform_feature_store_db_name
+  feature_store_server_name             = var.transform_feature_store_server_name
+  apps_ad_group_display_name            = var.transform_apps_ad_group_display_name
+  developers_ad_group_display_name      = var.transform_developers_ad_group_display_name
```
Comment (jjgriff93) [inline apps/main.tf:46]: `developers_ad_group_display_name` Do we use this? Doesn't look like we do

File: apps/app/webapp.tf (hunk 3 above)
Comment (jjgriff93) [inline apps/app/webapp.tf:220]: Also suggest you move this to the `cosmos.tf` file as it pertains to cosmos and not the web app

[the remark this answers is not in the document]
Comment (martinpeck) [reply apps/app/webapp.tf:220]: Honestly, not sure where that data will be stored right now. It's possible that it'll use CosmosDB, but not confirmed.

[the remark this answers is not in the document]
Comment (jedheaj314) [reply apps/app/webapp.tf]: for_each implemented

**Comment by t-young31** [comment]
I think https://github.com/UCLH-Foundry/FlowEHR/blob/5114fa38af4e4ec998ace3ec8f71af69449db70a/infrastructure/serve/cosmos.tf#L23-L24

should be dependent on `accesses_real_data` if the goal is to be able to access persistent environments from local(?)

**Comment by jedheaj314** [comment]
> @jedheaj314 just to confirm as mentioned on stand-up, have you tested running an app locally on your machine (the `Dash-Seedling` app), populating the connection string to your FlowEHR Cosmos DB, and tested connectivity via MSI? I'll give this a go as well, just wanted to check

Done 44ef954282d232a59d435b1904e7c1a45b921736

**Comment by jedheaj314** [comment]
> I think
> 
> https://github.com/UCLH-Foundry/FlowEHR/blob/5114fa38af4e4ec998ace3ec8f71af69449db70a/infrastructure/serve/cosmos.tf#L23-L24
> 
> should be dependent on `accesses_real_data` if the goal is to be able to access persistent environments from local(?)

Done 44ef954282d232a59d435b1904e7c1a45b921736

File: apps/variables.tf [hunk 7]
```diff
@@ -62,15 +62,15 @@ variable "transform_apps_ad_group_display_name" {
   type = string
 }
 
-variable "transform_developers_ad_group_display_name" {
+variable "transform_apps_ad_group_principal_id" {
```
Comment (t-young31) [inline apps/variables.tf:65]: named correctly? seems inconsistent with the others

Comment (jedheaj314) [reply apps/variables.tf:65]: Yes, as it comes from transform and also is consistent with other group related variables.
What would you recommend/name?

Comment (t-young31) [reply apps/variables.tf:65]: oh the AD groups for apps is created in transform? fair enough! I would have expected in in apps/ but 🤷🏼

Comment (jedheaj314) [reply apps/variables.tf:65]: Can move that in apps. Will do that in a separate PR.

**Review by t-young31 — APPROVED** [review]
LGTM pending @jjgriff93 's investigation into https://github.com/UCLH-Foundry/FlowEHR/blob/908e94fc8b1ae85e3c5fccc1da8d3e81192ee11c/infrastructure/serve/cosmos.tf#L43

File: apps/app/data.tf [hunk 8]
```diff
@@ -64,3 +64,9 @@ data "template_file" "core_github_workflow" {
     branch_name                = local.core_branch_name
   }
 }
+
+data "azurerm_cosmosdb_sql_role_definition" "data_contributor" {
+  resource_group_name = var.resource_group_name
+  account_name        = var.cosmos_account_name
+  role_definition_id  = "00000000-0000-0000-0000-000000000002" # Azure Cosmos Built in role ID
```
Comment (jjgriff93) [inline apps/app/data.tf:71]:
```suggestion
  role_definition_id  = "00000000-0000-0000-0000-000000000002" # Cosmos Data Contributor built-in role ID
```

File: apps/terragrunt.hcl [hunk 9]
```diff
@@ -124,20 +124,22 @@ dependency "serve" {
...
+  naming_suffix_truncated                    = dependency.core.outputs.naming_suffix_truncated
+  core_rg_name                               = dependency.core.outputs.core_rg_name
+  core_rg_location                           = dependency.core.outputs.core_rg_location
+  core_kv_id                                 = dependency.core.outputs.core_kv_id
+  core_log_analytics_name                    = dependency.core.outputs.core_log_analytics_name
+  serve_webapps_subnet_id                    = dependency.core.outputs.webapps_subnet_id
+  core_developers_ad_group_principal_id      = dependency.core.outputs.developers_ad_group_principal_id
+  core_data_scientists_ad_group_principal_id = dependency.core.outputs.data_scientists_ad_group_principal_id
+
+
```
Comment (jjgriff93) [inline apps/terragrunt.hcl:137]:
```suggestion
```

File: apps/terragrunt.hcl [hunk 10]
```diff
@@ -124,20 +124,22 @@ dependency "serve" {
...
+  transform_feature_store_server_name  = dependency.transform.outputs.feature_store_server_name
+  transform_feature_store_db_name      = dependency.transform.outputs.feature_store_db_name
+  transform_apps_ad_group_display_name = dependency.transform.outputs.apps_ad_group_display_name
+  transform_apps_ad_group_principal_id = dependency.transform.outputs.apps_ad_group_principal_id
+
```
Comment (jjgriff93) [inline apps/terragrunt.hcl:142]:
```suggestion
```

File: infrastructure/core/groups.tf [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  owners           = [data.azurerm_client_config.current.object_id]
+  security_enabled = true
+}
+
+resource "azuread_group" "ad_group_data_scientists" {
+  count            = var.accesses_real_data ? 0 : 1
+  display_name     = "${var.suffix_override} flowehr-data-scientists"
+  owners           = [data.azurerm_client_config.current.object_id]
+  security_enabled = true
+}
```
Comment (jjgriff93) [inline infrastructure/core/groups.tf:27]:
```suggestion
}

```

**Review by jjgriff93 — APPROVED** [review]
Good stuff!

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4718755771 (with refid `2c411d7c`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/218#issuecomment-1510869318) from @jedheaj314)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4719325555 (with refid `2c411d7c`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/218#issuecomment-1510960287) from @jedheaj314)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4719539375 (with refid `2c411d7c`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/218#issuecomment-1510995215) from @jedheaj314)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4720553708 (with refid `2c411d7c`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/218#issuecomment-1511156312) from @jedheaj314)

**Pull request merged**